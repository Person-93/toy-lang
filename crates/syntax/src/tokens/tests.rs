use super::*;
use anyhow::{Context, Result};
use insta::{assert_snapshot, with_settings};
use logos::{Logos, Span};
use std::{
  fmt::{self, Display, Formatter},
  fs,
};

#[test]
fn snapshot() -> Result<()> {
  let dir = concat!(env!("CARGO_MANIFEST_DIR"), "/../../toys");
  for entry in fs::read_dir(dir).with_context(|| format!("missing toys directory: {dir}"))? {
    let entry = entry.context("failed reading directory entry")?;
    if !entry
      .file_type()
      .context("failed reading file type")?
      .is_file()
    {
      continue;
    }
    let path = entry.path();
    if let Some(ext) = path.extension() {
      if ext.to_str().unwrap() != "toy" {
        continue;
      }
    }

    let text = fs::read_to_string(&path)
      .context("failed reading file")?
      // strip '\r' so snapshots match on windows
      .replace('\r', "");
    with_settings!({snapshot_suffix => path.file_name().unwrap().to_str().unwrap()}, {
      assert_snapshot!(Token::lexer(&text)
        .spanned()
        .map(|(token, span)| Output(token, span.clone(), &text[span]).to_string())
        .collect::<Vec<_>>()
        .join("\n"))
    });
  }

  return Ok(());

  struct Output<'a>(Token, Span, &'a str);

  impl Display for Output<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
      match &self.0 {
        Token::Error => write!(f, "INVALID @ {:?}: {}", self.1, self.2),
        Token::StrLit(lit) => write!(f, "StrLit @ {:?}: \"{lit}\"", self.1),
        Token::NumLit(lit) => write!(f, "NumLit @ {:?}: {lit}", self.1),
        Token::Ident(ident) => write!(f, "Ident @ {:?}: {ident}", self.1),
        _ => write!(f, "{:?} @ {:?}", self.0, self.1),
      }
    }
  }
}
