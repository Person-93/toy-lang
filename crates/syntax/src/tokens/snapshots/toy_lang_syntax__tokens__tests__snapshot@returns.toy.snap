---
source: crates/syntax/src/tokens/tests.rs
assertion_line: 34
expression: "Token::lexer(&text).spanned().map(|(token, span)|\n                                      Output(token, span.clone(),\n                                             &text[span]).to_string()).collect::<Vec<_>>().join(\"\\n\")"

---
KwFn @ 0..2
Ident @ 3..13: return_str
LeftParen @ 13..14
RightParen @ 14..15
Returns @ 16..18
Ampersand @ 19..20
Ident @ 20..23: str
Colon @ 23..24
LeftBrace @ 25..26
StrLit @ 27..40: "hello world"
RightBrace @ 41..42
KwFn @ 44..46
Ident @ 47..57: return_num
LeftParen @ 57..58
RightParen @ 58..59
Returns @ 60..62
Ident @ 63..65: i8
Colon @ 65..66
LeftBrace @ 67..68
NumLit @ 69..71: 42
RightBrace @ 72..73
